File: src/gamedatagen/core/leveling_system.py
"""
Leveling System for XP budgeting and content scaling

Ensures players can reach max level through available content.
"""

from typing import Dict, List, Optional

from pydantic import BaseModel


class XPCurve(BaseModel):
    """XP curve configuration"""

    curve_type: str = "exponential"  # exponential, linear, logarithmic
    base_xp: int = 100
    exponent: float = 1.5
    level_multiplier: float = 1.0


class ContentBudget(BaseModel):
    """Content distribution for XP budgeting"""

    main_quests_percent: float = 0.35
    side_quests_percent: float = 0.25
    dungeons_percent: float = 0.20
    events_percent: float = 0.10
    grinding_percent: float = 0.10


class LevelingSystem:
    """XP budgeting and content scaling system"""

    def __init__(
        self,
        max_level: int = 100,
        xp_curve: Optional[XPCurve] = None,
        content_budget: Optional[ContentBudget] = None,
    ) -> None:
        self.max_level = max_level
        self.xp_curve = xp_curve or XPCurve()
        self.content_budget = content_budget or ContentBudget()

        # Calculate XP per level
        self.xp_per_level = self._calculate_xp_table()
        self.total_xp_needed = sum(self.xp_per_level.values())

    def _calculate_xp_table(self) -> Dict[int, int]:
        """Calculate XP needed for each level"""
        xp_table = {}

        if self.xp_curve.curve_type == "exponential":
            curve_fn = self._exponential_curve
        elif self.xp_curve.curve_type == "linear":
            curve_fn = self._linear_curve
        elif self.xp_curve.curve_type == "logarithmic":
            curve_fn = self._logarithmic_curve
        else:
            curve_fn = self._exponential_curve

        for level in range(1, self.max_level + 1):
            xp_table[level] = curve_fn(level)

        return xp_table

    def _exponential_curve(self, level: int) -> int:
        """Exponential XP curve (gets harder as you level)"""
        return int(
            self.xp_curve.base_xp
            * (level ** self.xp_curve.exponent)
            * self.xp_curve.level_multiplier
        )

    def _linear_curve(self, level: int) -> int:
        """Linear XP curve (constant increase)"""
        return int(self.xp_curve.base_xp * level * self.xp_curve.level_multiplier)

    def _logarithmic_curve(self, level: int) -> int:
        """Logarithmic XP curve (easier at higher levels)"""
        import math

        return int(
            self.xp_curve.base_xp
            * math.log(level + 1)
            * level
            * self.xp_curve.level_multiplier
        )

    def get_level_range_xp(self, min_level: int, max_level: int) -> int:
        """Get total XP needed for a level range"""
        return sum(self.xp_per_level[level] for level in range(min_level, max_level + 1))

    def calculate_content_requirements(
        self,
    ) -> Dict[str, Dict[str, int]]:
        """
        Calculate how much content is needed to reach max level

        Returns budget breakdown per content type.
        """
        budget = {}

        # Main quests
        main_quest_xp = int(self.total_xp_needed * self.content_budget.main_quests_percent)
        avg_main_quest_xp = 500  # Configurable
        budget["main_quests"] = {
            "total_xp": main_quest_xp,
            "count": main_quest_xp // avg_main_quest_xp,
            "avg_xp": avg_main_quest_xp,
        }

        # Side quests
        side_quest_xp = int(self.total_xp_needed * self.content_budget.side_quests_percent)
        avg_side_quest_xp = 200
        budget["side_quests"] = {
            "total_xp": side_quest_xp,
            "count": side_quest_xp // avg_side_quest_xp,
            "avg_xp": avg_side_quest_xp,
        }

        # Dungeons
        dungeon_xp = int(self.total_xp_needed * self.content_budget.dungeons_percent)
        avg_dungeon_xp = 1000
        budget["dungeons"] = {
            "total_xp": dungeon_xp,
            "count": dungeon_xp // avg_dungeon_xp,
            "avg_xp": avg_dungeon_xp,
        }

        # Events
        event_xp = int(self.total_xp_needed * self.content_budget.events_percent)
        avg_event_xp = 300
        budget["events"] = {
            "total_xp": event_xp,
            "count": event_xp // avg_event_xp,
            "avg_xp": avg_event_xp,
        }

        return budget

    def validate_content_xp(
        self,
        main_quest_count: int,
        side_quest_count: int,
        dungeon_count: int,
        event_count: int,
        avg_quest_level: Optional[int] = None,
    ) -> Dict[str, any]:
        """
        Validate if available content provides enough XP

        Returns:
            {
                "valid": bool,
                "total_xp": int,
                "deficit": int,  # Negative if not enough XP
                "suggestions": List[str]
            }
        """
        requirements = self.calculate_content_requirements()

        # Calculate available XP
        available_xp = (
            main_quest_count * requirements["main_quests"]["avg_xp"]
            + side_quest_count * requirements["side_quests"]["avg_xp"]
            + dungeon_count * requirements["dungeons"]["avg_xp"]
            + event_count * requirements["events"]["avg_xp"]
        )

        deficit = available_xp - self.total_xp_needed
        valid = deficit >= 0

        suggestions = []
        if not valid:
            # Calculate how much more content is needed
            needed_xp = abs(deficit)

            suggestions.append(
                f"Need {needed_xp:,} more XP to reach level {self.max_level}"
            )

            # Suggest content to add
            main_quests_needed = needed_xp // requirements["main_quests"]["avg_xp"]
            if main_quests_needed > 0:
                suggestions.append(f"Add ~{main_quests_needed} main quests")

            side_quests_needed = needed_xp // requirements["side_quests"]["avg_xp"]
            if side_quests_needed > 0:
                suggestions.append(f"Or add ~{side_quests_needed} side quests")

        return {
            "valid": valid,
            "total_xp": available_xp,
            "needed_xp": self.total_xp_needed,
            "deficit": deficit,
            "suggestions": suggestions,
        }

    def scale_quest_xp(self, quest_level: int, base_xp: int = 100) -> int:
        """Scale quest XP based on level"""
        # XP reward scales with level
        return int(base_xp * (1 + (quest_level - 1) * 0.1))

    def distribute_content_across_levels(
        self, content_count: int, min_level: int = 1, max_level: Optional[int] = None
    ) -> List[int]:
        """
        Distribute content across level range

        Returns list of levels for each piece of content.
        """
        if max_level is None:
            max_level = self.max_level

        levels = []
        level_range = max_level - min_level + 1

        # Distribute evenly across levels
        for i in range(content_count):
            # Round-robin distribution
            level = min_level + (i % level_range)
            levels.append(level)

        return sorted(levels)

    def get_stats(self) -> Dict[str, any]:
        """Get leveling system statistics"""
        requirements = self.calculate_content_requirements()

        return {
            "max_level": self.max_level,
            "total_xp_needed": self.total_xp_needed,
            "xp_curve": self.xp_curve.curve_type,
            "content_requirements": requirements,
            "avg_xp_per_level": self.total_xp_needed // self.max_level,
        }

File: src/core/LevelingSystem.ts
/**
 * Leveling system for managing XP budgeting and level scaling
 * Ensures players reach max level through available content
 */

export interface LevelingConfig {
  maxLevel: number;
  baseXP: number; // XP needed for level 2
  xpCurve: 'linear' | 'exponential' | 'logarithmic' | 'custom';
  customCurveMultiplier?: number; // For custom curves
}

export interface ContentXPBudget {
  mainQuests: number; // % of total XP
  sideQuests: number;
  dungeons: number;
  raids: number;
  gathering: number;
  crafting: number;
  exploration: number;
  worldBosses: number;
  dailies: number;
  other: number;
}

export interface LevelRange {
  min: number;
  max: number;
}

export interface QuestXPReward {
  questLevel: number;
  baseXP: number;
  bonusXP?: number;
}

/**
 * Leveling system for XP budgeting and content scaling
 */
export class LevelingSystem {
  private config: LevelingConfig;
  private xpTable: Map<number, number> = new Map(); // Level → Total XP needed
  private budget: ContentXPBudget;

  constructor(
    config: LevelingConfig,
    budget: ContentXPBudget = {
      mainQuests: 0.4,
      sideQuests: 0.15,
      dungeons: 0.2,
      raids: 0.05,
      gathering: 0.05,
      crafting: 0.03,
      exploration: 0.05,
      worldBosses: 0.04,
      dailies: 0.02,
      other: 0.01,
    }
  ) {
    this.config = config;
    this.budget = budget;
    this.generateXPTable();
  }

  /**
   * Generate XP table for all levels
   */
  private generateXPTable(): void {
    this.xpTable.set(1, 0); // Level 1 starts at 0 XP

    for (let level = 2; level <= this.config.maxLevel; level++) {
      const xpForLevel = this.calculateXPForLevel(level);
      const previousTotal = this.xpTable.get(level - 1) || 0;
      this.xpTable.set(level, previousTotal + xpForLevel);
    }
  }

  /**
   * Calculate XP needed to go from level N to level N+1
   */
  private calculateXPForLevel(level: number): number {
    const base = this.config.baseXP;

    switch (this.config.xpCurve) {
      case 'linear':
        return base * level;

      case 'exponential':
        return Math.floor(base * Math.pow(level, 1.5));

      case 'logarithmic':
        return Math.floor(base * (level + Math.log(level + 1) * 10));

      case 'custom':
        const multiplier = this.config.customCurveMultiplier || 1.2;
        return Math.floor(base * Math.pow(level, multiplier));

      default:
        return base * level;
    }
  }

  /**
   * Get total XP needed to reach a specific level
   */
  getTotalXPForLevel(level: number): number {
    return this.xpTable.get(level) || 0;
  }

  /**
   * Get XP needed to go from one level to another
   */
  getXPBetweenLevels(fromLevel: number, toLevel: number): number {
    return this.getTotalXPForLevel(toLevel) - this.getTotalXPForLevel(fromLevel);
  }

  /**
   * Get total XP available in the game
   */
  getTotalXPNeeded(): number {
    return this.getTotalXPForLevel(this.config.maxLevel);
  }

  /**
   * Calculate XP budget for a content type
   */
  getXPBudget(contentType: keyof ContentXPBudget): number {
    const totalXP = this.getTotalXPNeeded();
    const percentage = this.budget[contentType];
    return Math.floor(totalXP * percentage);
  }

  /**
   * Calculate how many pieces of content needed to fill XP budget
   */
  calculateContentCount(
    contentType: keyof ContentXPBudget,
    averageXPPerPiece: number
  ): number {
    const budget = this.getXPBudget(contentType);
    return Math.ceil(budget / averageXPPerPiece);
  }

  /**
   * Generate quest XP reward based on quest level
   */
  calculateQuestXP(
    questLevel: number,
    questType: 'main' | 'side' | 'daily' = 'side',
    difficulty: 'easy' | 'normal' | 'hard' = 'normal'
  ): number {
    const baseXP = this.calculateXPForLevel(questLevel);

    let multiplier = 1.0;

    // Quest type multipliers
    switch (questType) {
      case 'main':
        multiplier = 1.5;
        break;
      case 'side':
        multiplier = 1.0;
        break;
      case 'daily':
        multiplier = 0.8;
        break;
    }

    // Difficulty multipliers
    switch (difficulty) {
      case 'easy':
        multiplier *= 0.8;
        break;
      case 'normal':
        multiplier *= 1.0;
        break;
      case 'hard':
        multiplier *= 1.3;
        break;
    }

    return Math.floor(baseXP * multiplier);
  }

  /**
   * Calculate dungeon XP reward
   */
  calculateDungeonXP(
    dungeonLevel: number,
    type: 'normal' | 'heroic' | 'mythic' = 'normal',
    bossCount: number = 3
  ): number {
    const baseXP = this.calculateXPForLevel(dungeonLevel);

    let multiplier = bossCount * 2; // Base multiplier per boss

    switch (type) {
      case 'normal':
        multiplier *= 1.0;
        break;
      case 'heroic':
        multiplier *= 1.5;
        break;
      case 'mythic':
        multiplier *= 2.0;
        break;
    }

    return Math.floor(baseXP * multiplier);
  }

  /**
   * Suggest level ranges for zones based on max level
   */
  suggestZoneLevelRanges(zoneCount: number): LevelRange[] {
    const maxLevel = this.config.maxLevel;
    const levelsPerZone = Math.floor(maxLevel / zoneCount);
    const ranges: LevelRange[] = [];

    for (let i = 0; i < zoneCount; i++) {
      const min = i * levelsPerZone + 1;
      const max = Math.min((i + 1) * levelsPerZone, maxLevel);
      ranges.push({ min, max });
    }

    // Adjust last zone to include max level
    if (ranges.length > 0) {
      ranges[ranges.length - 1].max = maxLevel;
    }

    return ranges;
  }

  /**
   * Distribute content across level ranges
   */
  distributeContent(
    totalCount: number,
    levelRanges: LevelRange[]
  ): Map<LevelRange, number> {
    const distribution = new Map<LevelRange, number>();

    // Simple even distribution
    const baseCount = Math.floor(totalCount / levelRanges.length);
    const remainder = totalCount % levelRanges.length;

    for (let i = 0; i < levelRanges.length; i++) {
      const count = baseCount + (i < remainder ? 1 : 0);
      distribution.set(levelRanges[i], count);
    }

    return distribution;
  }

  /**
   * Calculate if player can reach max level with available content
   */
  validateXPBudget(availableContent: {
    mainQuestCount: number;
    sideQuestCount: number;
    dungeonCount: number;
    averageQuestLevel: number;
    averageDungeonLevel: number;
  }): {
    valid: boolean;
    totalAvailableXP: number;
    totalRequiredXP: number;
    deficit: number;
    suggestions: string[];
  } {
    const totalRequired = this.getTotalXPNeeded();

    let totalAvailable = 0;

    // Main quests
    totalAvailable +=
      availableContent.mainQuestCount *
      this.calculateQuestXP(availableContent.averageQuestLevel, 'main');

    // Side quests
    totalAvailable +=
      availableContent.sideQuestCount *
      this.calculateQuestXP(availableContent.averageQuestLevel, 'side');

    // Dungeons
    totalAvailable +=
      availableContent.dungeonCount *
      this.calculateDungeonXP(availableContent.averageDungeonLevel);

    const deficit = totalRequired - totalAvailable;
    const suggestions: string[] = [];

    if (deficit > 0) {
      // Suggest additional content
      const additionalMainQuests = Math.ceil(
        deficit /
          this.calculateQuestXP(availableContent.averageQuestLevel, 'main')
      );
      suggestions.push(
        `Add ${additionalMainQuests} more main quests to fill XP gap`
      );

      const additionalDungeons = Math.ceil(
        deficit / this.calculateDungeonXP(availableContent.averageDungeonLevel)
      );
      suggestions.push(
        `Or add ${additionalDungeons} more dungeons to fill XP gap`
      );
    }

    return {
      valid: deficit <= 0,
      totalAvailableXP: totalAvailable,
      totalRequiredXP: totalRequired,
      deficit: Math.max(0, deficit),
      suggestions,
    };
  }

  /**
   * Get XP table as array
   */
  getXPTable(): Array<{ level: number; totalXP: number; xpToNext: number }> {
    const table: Array<{ level: number; totalXP: number; xpToNext: number }> =
      [];

    for (let level = 1; level <= this.config.maxLevel; level++) {
      const totalXP = this.getTotalXPForLevel(level);
      const xpToNext =
        level < this.config.maxLevel
          ? this.calculateXPForLevel(level + 1)
          : 0;

      table.push({ level, totalXP, xpToNext });
    }

    return table;
  }

  /**
   * Update budget allocation
   */
  updateBudget(budget: Partial<ContentXPBudget>): void {
    this.budget = { ...this.budget, ...budget };

    // Validate budget sums to ~1.0
    const total = Object.values(this.budget).reduce((sum, val) => sum + val, 0);
    if (Math.abs(total - 1.0) > 0.01) {
      console.warn(
        `XP budget does not sum to 1.0 (current: ${total.toFixed(2)})`
      );
    }
  }

  /**
   * Get current config
   */
  getConfig(): LevelingConfig {
    return { ...this.config };
  }

  /**
   * Get current budget
   */
  getBudget(): ContentXPBudget {
    return { ...this.budget };
  }
}

// Global leveling system instance (will be initialized by GameDataGen)
export let globalLevelingSystem: LevelingSystem | null = null;

/**
 * Initialize global leveling system
 */
export function initializeLevelingSystem(
  config: LevelingConfig,
  budget?: ContentXPBudget
): LevelingSystem {
  globalLevelingSystem = new LevelingSystem(config, budget);
  return globalLevelingSystem;
}
